### PR TITLE
Noisy Actions & Noise-Masked Observations

### PR DESCRIPTION
**Pull request by decodyng**
**Primary Experimental Features** 
- Modifies AdversaryMasking code to allow for masking observations with noise as well as zero or initialization values 
- Uses NoisyAgentWrapper to allow for testing how adversary and zoo policies respond to gradual application of noise 
- MultiScore and Experiment code to allow both of the above experiments to be run 
- Plotting code for visualization of masked observations results 

**Related Changes**
- In multi/score_worker, add the ability to pass in index keys for adding into ray results index
- Added separate ray config file for running of above experiments, and brought aws.yaml back into accordance with master, modulo changes for making a python wheel of modelfree/aprl

### CODE REVIEW

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env bash
+
+function wait_proc {
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: This is used in three files now (this, `attack_transfer.sh` and `baselines.sh`). I think time to refactor it into a separate file. For example, if we put `wait_proc` in a file `common.sh` then we could include that from other scripts by doing:
```bash
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
. ${DIR}/common.sh
```

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    fi
+}
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+OUT_ROOT=data/aws/score_agents
+TIMESTAMP=`date --iso-8601=seconds`
+
+MULTI_SCORE_CMD="python -m modelfree.multi.score with medium_accuracy "
+
+# Not sure what this is doing
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: Kill lines 19-23. You don't use log directories in this script. (FYI, the if statement is checking if there are no CLI arguments, in which case it exits abnormally printing a usage message.)

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+# Make a directory for each of the experiments we'll be running, to store results in
+for dir in noisy_adversary_actions noisy_victim_actions noisy_victim_obs; do
+    mkdir -p ${OUT_ROOT}/${dir}/${TIMESTAMP}
+done
+# Rerun highest_win_rate and store the results in ADVERSARY_PATHS, which we
+# export as an environment variable which multi score will use
+ADVERSARY_PATHS=${OUT_ROOT}/normal/2019-05-05T18:12:24+00:00/best_adversaries.json
+
+export ADVERSARY_PATHS=${ADVERSARY_PATHS}
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: You can do this in one line: `export ADVERSARY_PATHS=${OUT_ROOT}/normal/2019-05-05T18:12:24+00:00/best_adversaries.json`

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+if [[ $# -eq 0 ]]; then
+    echo "usage: $0 <logdir> [logdir ...]"
+    exit 1
+fi
+
+# Make a directory for each of the experiments we'll be running, to store results in
+for dir in noisy_adversary_actions noisy_victim_actions noisy_victim_obs; do
+    mkdir -p ${OUT_ROOT}/${dir}/${TIMESTAMP}
+done
+# Rerun highest_win_rate and store the results in ADVERSARY_PATHS, which we
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: This comment seems out of sync with the code - delete?

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+${MULTI_SCORE_CMD} zoo_baseline mask_observations_of_victim mask_observations_with_additive_noise \
+    save_path=${OUT_ROOT}/noisy_victim_obs/${TIMESTAMP}/noisy_zoo_observations.json&
+wait_proc
+echo "Additive noise masking zoo baseline complete"
+
+${MULTI_SCORE_CMD} adversary_trained mask_observations_of_victim mask_observations_with_additive_noise \
+    save_path=${OUT_ROOT}/noisy_victim_obs/${TIMESTAMP}/noisy_adversary_observations.json&
+wait_proc
+
+${MULTI_SCORE_CMD} zoo_baseline mask_observations_of_victim mask_observations_with_smaller_additive_noise \
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: This has the same save path as lines 54-57; won't they get clobbered?

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    save_path=${OUT_ROOT}/noisy_victim_obs/${TIMESTAMP}/noisy_zoo_observations.json&
+wait_proc
+echo "Additive noise masking zoo baseline complete"
+
+${MULTI_SCORE_CMD} adversary_trained mask_observations_of_victim mask_observations_with_smaller_additive_noise \
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: This has the same save path as lines 59-61; won't they get clobbered?

File: experiments/modelfree/noisy_actions_and_obs.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    mkdir -p ${OUT_ROOT}/${dir}/${TIMESTAMP}
+done
+# Rerun highest_win_rate and store the results in ADVERSARY_PATHS, which we
+# export as an environment variable which multi score will use
+ADVERSARY_PATHS=${OUT_ROOT}/normal/2019-05-05T18:12:24+00:00/best_adversaries.json
+
+export ADVERSARY_PATHS=${ADVERSARY_PATHS}
+
+
+${MULTI_SCORE_CMD} zoo_baseline noise_adversary_actions \
```
Comment (AdamGleave) [inline experiments/modelfree/noisy_actions_and_obs.sh]: This list of commands is very repetitive and I thought it hard to parse what experiments it was actually running.

I'd suggest simplifying by defining a function that abstracts some of the repeated elements. It seems the only things that change between lines are:
  - Experiment to run: `zoo_baseline` or `adversary_trained`.
  - Additional named configs: e.g. `noise_adversary_actions`.
  - Save path directory: e.g. `noisy_adversary_actions` or `noisy_victim_actions`.
  - Save path filename: e.g. `noisy_victim.json`, `noisy_adversary.json`.

I think we can simplify further by exploiting some patterns. For example, having a consistent naming convention between the paths and the named configs would let us only specify one of them. There are also some cases where we're taking all possible combinations of particular values, in which case a nested for loop will simplify things.

File: experiments/modelfree/plot_noisy_observations.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
+import json
+import math
+import os
+
+from matplotlib import pyplot as plt
+import pandas as pd
+import seaborn as sns
+
+RESULT_DIR = "/Users/cody/Data/adversarial_policies/ray_results/noisy_victim_obs/"
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Please don't use absolute directories :( We should aim to make the code run out of the box on anyone's machine.

The convention I've followed is to assume the CWD is the root of the project, and that the `data` subdirectory contains all results. If you upload your results to S3, then you can assume they exist in `data/aws/` -- which they will if someone has run the AWS sync script, and is easy for them to do if not.

Also I'm confused why the data lives in `ray_results`? That is just a temporary directory, and it should always be copied out of there at the end of the experiment.

File: experiments/modelfree/plot_noisy_observations.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+OUT_DIR = "/Users/cody/Data/adversarial_policies/noisy_obs_plots/results_may_22"
+env_lookup = {
+    'SumoHumans': 'multicomp/SumoHumansAutoContact-v0',
+    'SumoAnts': 'multicomp/SumoAntsAutoContact-v0',
+    'KickAndDefend': 'multicomp/KickAndDefend-v0',
+    # 'YouShallNotPass': 'multicomp/YouShallNotPassHumans-v0' (Removed for now because I don't have
+    # time at the moment to fix the plotting logic to deal with victim in different location)
+}
+available_zoos = {
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Duplicate of `modelfree.envs.gym_compete.NUM_ZOO_POLICIES`

Comment (AdamGleave) [reply experiments/modelfree/plot_noisy_observations.py]: Also, constants should be in all-caps: https://www.python.org/dev/peps/pep-0008/#constants

File: experiments/modelfree/plot_noisy_observations.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+import math
+import os
+
+from matplotlib import pyplot as plt
+import pandas as pd
+import seaborn as sns
+
+RESULT_DIR = "/Users/cody/Data/adversarial_policies/ray_results/noisy_victim_obs/"
+OUT_DIR = "/Users/cody/Data/adversarial_policies/noisy_obs_plots/results_may_22"
+env_lookup = {
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Consider using `modelfree.visualize.util.PRETTY_ENV` instead. (The key-values are the other way around, but it's easy to reverse if you need.)

File: experiments/modelfree/plot_noisy_observations.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    # 'YouShallNotPass': 'multicomp/YouShallNotPassHumans-v0' (Removed for now because I don't have
+    # time at the moment to fix the plotting logic to deal with victim in different location)
+}
+available_zoos = {
+    'SumoHumans': 3,
+    'SumoAnts': 4,
+    'KickAndDefend': 3
+}
+
+k_keys = ['env', 'agent0_type', 'agent0_path', 'agent1_type',
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Constants should be in all-caps: https://www.python.org/dev/peps/pep-0008/#constants

File: experiments/modelfree/plot_noisy_observations.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
+import json
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Should this live in `experiments/modelfree/` or in e.g. `src/modelfree/visualize/`? I've largely been using `experiments/` as a directory for ad-hoc, stand-alone scripts. I think could make a case for either here, but I think leaning towards including it in the main `src` directory, especially if we integrate it more with the rest of `modelfree`.

File: experiments/modelfree/plot_noisy_observations.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    return outp
+
+
+def noisy_adversary_opponent_subset_plot(original_df, subset_specs, transform_specs,
+                                         logistic=True, plot_line=True, savefile=None):
+    subset_df = subset(original_df, subset_specs)
+    if len(subset_df) == 0:
+        return
+    transformed_df = transform(subset_df, transform_specs)
+    plt.figure(figsize=(10, 7))
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Nit: using `plt` is error-prone as it defaults to the latest figure. Good practice is to use `fig = plt.figure`, and then the corresponding `Figure` instance methods: `fig.suptitle`, `fig.savefig`, `fig.show`, `fig.close`. (You don't need to change this here but I'd recommend it for new code, especially if you have any complicated nested plotting.)

File: experiments/modelfree/plot_noisy_observations.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            noisy_multiple_opponent_subset_plot(noisy_zoo_obs_df, subset_specs=subset_params,
+                                                transform_specs=baseline_transformations,
+                                                savefile=zoo_plot_path)
+            noisy_adversary_opponent_subset_plot(noisy_adv_obs_df, subset_specs=subset_params,
+                                                 transform_specs=baseline_transformations,
+                                                 savefile=os.path.join(adversary_plot_path))
+
+
+if __name__ == "__main__":
+    for experiment_run in ["ep_500_5-22_single_zoo", "ep_100_5-21", "ep_500_5-22_all_zoo"]:
```
Comment (AdamGleave) [inline experiments/modelfree/plot_noisy_observations.py]: Will the path names always be these?

File: src/modelfree/configs/multi/score.py [hunk 15]
```diff
@@ -111,6 +112,15 @@ def high_accuracy(score):
         score['episodes'] = 1000
         score['num_env'] = 16
 
+    @multi_score_ex.named_config
+    def medium_accuracy(score):
+        score = dict(score)
+        score['episodes'] = 100
+        score['num_env'] = 16
+        exp_name = 'medium_accuracy_'
```
Comment (AdamGleave) [inline src/modelfree/configs/multi/score.py]: Adding a tag seems good. But won't this just override the experiment name? I think this needs to be `'medium_acuracy_' + exp_name`. For consistency should probably also update `high_accuracy` to add a tag similarly.

File: src/modelfree/configs/multi/score.py [hunk 16]
```diff
@@ -165,6 +175,59 @@ def mask_observations_with_zeros(exp_name, score):
         score['mask_agent_kwargs'] = {'masking_type': 'zeros'}
         exp_name = 'zero_' + exp_name
 
+    @multi_score_ex.named_config
+    def mask_observations_with_additive_noise(exp_name, score, spec):
+        score = score.copy()
```
Comment (AdamGleave) [inline src/modelfree/configs/multi/score.py]: Why are we using `score = score.copy()` here rather than `score = dict(score)`?

I don't particularly care which we use, but we should be consistent throughout. If there's a good reason for using `score.copy()` lets change all of them to that; otherwise keep it as `dict(score)`?

File: src/modelfree/configs/multi/score.py [hunk 17]
```diff
@@ -165,6 +175,59 @@ def mask_observations_with_zeros(exp_name, score):
...
+        score['index_keys'] = ['hashable_mask_agent_kwargs', 'mask_agent_noise']
+        score['mask_agent_kwargs'] = {
+            'masking_type': 'additive_noise'
+        }
+        spec['num_samples'] = 25
+        spec['config']['mask_agent_noise'] = tune.sample_from(
+            lambda spec: np.random.exponential(scale=1.0)
+        )
+        exp_name = 'additive_noise_' + exp_name
```
Comment (AdamGleave) [inline src/modelfree/configs/multi/score.py]: Consider different experiment name than above?

File: src/modelfree/configs/multi/score.py [hunk 18]
```diff
@@ -165,6 +175,59 @@ def mask_observations_with_zeros(exp_name, score):
         score['mask_agent_kwargs'] = {'masking_type': 'zeros'}
         exp_name = 'zero_' + exp_name
 
+    @multi_score_ex.named_config
+    def mask_observations_with_additive_noise(exp_name, score, spec):
+        score = score.copy()
+        score['index_keys'] = ['hashable_mask_agent_kwargs', 'mask_agent_noise']
```
Comment (AdamGleave) [inline src/modelfree/configs/multi/score.py]: A lot of code duplicated between this and the subsequent function; factor out into a helper method? (Basically all the part that sets `score`.)

File: src/modelfree/configs/ray/aws.yaml [hunk 19]
```diff
@@ -2,19 +2,20 @@ cluster_name: aprl
 min_workers: 0
 max_workers: 10
 initial_workers: 0
-target_utilization_fraction: 0.8
+target_utilization_fraction: 0.4
```
Comment (AdamGleave) [inline src/modelfree/configs/ray/aws.yaml]: Why do we want our cluster to be 60% idle?

File: src/modelfree/configs/ray/aws.yaml [hunk 20]
```diff
@@ -2,19 +2,20 @@ cluster_name: aprl
...
 idle_timeout_minutes: 5
 
 docker:
   image: humancompatibleai/adversarial_policies:latest
-  container_name: ray
+  container_name: ray-cody
```
Comment (AdamGleave) [inline src/modelfree/configs/ray/aws.yaml]: Good practice to suffix the container name with your username but this should not be committed.

I make a copy of the file, `aws-localcfg.yaml`, which is gitignore'd, and make these changes. Then periodically do a manual diff between them and copy over any changes that are important to `aws.yaml`.

File: src/modelfree/configs/ray/aws-noisy-actions-observations.yaml [hunk 21]
```diff
@@ -0,0 +1,103 @@
+cluster_name: aprl-cody
```
Comment (AdamGleave) [inline src/modelfree/configs/ray/aws-noisy-actions-observations.yaml]: I'd be sad if we had to add a whole new Ray config file just for this use case :(

It seems some of the differences aren't needed:
  - Docker image. Not necessary once changes are merged.
  - Instance type. Not sure why this changed TBH.
  - Checking out the `noisy_adversary_actions` branch, also not necessary once merged.

The differences I can see that are needed are:
  - IamInstanceProfile. I think I'm ok adding this to the main `aws.yaml`. It does make it slightly harder to set AWS up, but I can see access to S3 being needed for a bunch of applications.
  - Syncing the data. Not OK with this being added, but maybe we can factor it out to a bash script in experiments/. Then a user who is running these experiments can just make a one-line change to the Ray config?

File: src/modelfree/envs/observation_masking.py [hunk 22]
```diff
@@ -11,49 +11,55 @@ def make_mask_from_class(cls):
...
-                        outp[other_agent] = np.full(self.other_agent_shapes[other_agent],
-                                                    fill_value=-4.2)
-                    elif self.masking_type == 'initialization':
-                        outp[other_agent] = self.initial_values[other_agent]
-                    else:
-                        raise ValueError(f"Unsupported masking type '{self.masking_type}'")
-            return outp
-
-        def get_other_qpos(self):
+        def _get_masking_given_initial(self, initial_position_value, true_current_position):
```
Comment (AdamGleave) [inline src/modelfree/envs/observation_masking.py]: :+1: this is much cleaner

File: src/modelfree/score_agent.py [hunk 23]
```diff
@@ -166,14 +168,18 @@ def default_score_config():
...
 
+    noisy_agent_index = None              # Agent to add add action noise to
+    noisy_agent_magnitude = 1.0           # Magnitude of action noise for agent being noised
     mask_agent_index = None               # index of agent whose observations should be limited
     mask_agent_kwargs = {                 # control how agent observations are limited
         'masking_type': 'initialization',
     }
-
+    mask_agent_noise = None
+    hashable_mask_agent_kwargs = '_'.join([k + '_' + v for k, v in mask_agent_kwargs.items()])
```
Comment (AdamGleave) [inline src/modelfree/score_agent.py]: Is this needed just because it's not a top-level config parameter? Can we just make `masking_type` top-level? This feels very confusing.

File: src/modelfree/score_agent.py [hunk 24]
```diff
@@ -18,6 +18,7 @@
 from modelfree.common.utils import TrajectoryRecorder, VideoWrapper, make_env, simulate
 from modelfree.envs.gym_compete import GymCompeteToOurs, game_outcome
 from modelfree.envs.observation_masking import make_mask_agent_wrappers
+from modelfree.training.shaping_wrappers import NoisyAgentWrapper
```
Comment (AdamGleave) [inline src/modelfree/score_agent.py]: Only `modelfree.train` should depend on `modelfree.training`. Let's move `NoisyAgentWrapper` somewhere into `common`. Probably `common.utils`, although that's becoming a bit of a kitchen sink. `utils.py` already has things for `DummyModel`, etc -- so we could maybe factor all those out into a new module e.g. `common.policies`? And leave things like `VideoWrapper`, `TrajectoryRecorder`, etc.

File: experiments/modelfree/noisy_actions_and_obs.sh (hunk 2 above)
Comment (decodyng) [reply experiments/modelfree/noisy_actions_and_obs.sh]: Sounds good. This was very much a Chesterton's Fence part of the script for me.

File: experiments/modelfree/noisy_actions_and_obs.sh (hunk 5 above)
Comment (decodyng) [reply experiments/modelfree/noisy_actions_and_obs.sh]: Ah, you're right, I never ran both of these at once because the smaller noise was a later add-on after the bigger noise was commented out. Will fix.

File: experiments/modelfree/plot_noisy_observations.py (hunk 8 above)
Comment (decodyng) [reply experiments/modelfree/plot_noisy_observations.py]: ray_results is just a directory elsewhere on my machine, not the temporary one you're thinking of. 

It generally felt more fault-tolerant to me to store results elsewhere on my machine rather than trying to reason about the whole complex set of code expectations about what was and wasn't expected to be in /data/aws at any given time, but I can refactor it.

File: experiments/modelfree/plot_noisy_observations.py (hunk 12 above)
Comment (decodyng) [reply experiments/modelfree/plot_noisy_observations.py]: Yeah, I didn't really know what your internal heuristic for that here, but it did seem more like post-experiment analysis than actual source code used in an experiment, which was my logic for putting it here.

File: src/modelfree/configs/multi/score.py (hunk 15 above)
Comment (decodyng) [reply src/modelfree/configs/multi/score.py]: My recollection was that exp_name wasn't defined at this point because of the order of how named_configs were run, but I can go back and try to confirm that.

File: src/modelfree/configs/multi/score.py (hunk 16 above)
Comment (decodyng) [reply src/modelfree/configs/multi/score.py]: Okay

File: src/modelfree/configs/multi/score.py (hunk 18 above)
Comment (decodyng) [reply src/modelfree/configs/multi/score.py]: I'm pretty hesitant to add another level of factored-out complexity given how complex these dynamic ray/sacred configurations already are to reason about.

File: src/modelfree/configs/ray/aws.yaml (hunk 19 above)
Comment (decodyng) [reply src/modelfree/configs/ray/aws.yaml]: This was from when I thought the problem with number of CPUs not going up was that it somehow was just under the utilization fraction, and I was putting this down to see if there'd be new workers created. Meant to revert it, just forgot to.

File: src/modelfree/configs/ray/aws.yaml (hunk 20 above)
Comment (decodyng) [reply src/modelfree/configs/ray/aws.yaml]: Okay

File: src/modelfree/configs/ray/aws-noisy-actions-observations.yaml (hunk 21 above)
Comment (decodyng) [reply src/modelfree/configs/ray/aws-noisy-actions-observations.yaml]: I put it in its own file so we could actually have a configuration that was known to work to run these experiments, rather than e.g. having to remember somewhere that it was necessary to add the data syncing commands in order to make that particular set of experiments work. Given that this was so fiddly to get to work in this first place, it seems worrying to lose something that out-of-the-box works for that code just for the sake of non-duplication. 

Instance type was an accident to leave in, I think.

File: src/modelfree/multi/common.py [hunk 25]
```diff
@@ -126,7 +126,7 @@ def local_config(platform, local_dir, spec):
             platform = 'local'
 
         if platform == 'local':
-            spec['sync_function'] = 'mkdir -p {remote_dir} && cp -aTv {local_dir}/ {remote_dir}'
+            spec['sync_function'] = 'mkdir -p {remote_dir} && cp -aTv {local_dir}/* {remote_dir}'
```
[the remark this answers is not in the document]
Comment (decodyng) [reply src/modelfree/multi/common.py]: Sure

File: src/modelfree/multi/score_worker.py [hunk 26]
```diff
@@ -29,6 +29,9 @@ def score_worker(base_config, tune_config, reporter):
     observer = FileStorageObserver.create(osp.join('data', 'sacred', 'score'))
     score_ex.observers.append(observer)
     run = score_ex.run(config_updates=config)
+    index_keys = config.get("index_keys", [])
+
+    idx = {k: v for k, v in config.items()
```
[the remark this answers is not in the document]
Comment (decodyng) [reply src/modelfree/multi/score_worker.py]: The logic of keeping it was to make sure that the required keys would be kept regardless of what was passed in to index_keys. (Which I guess it actually functioning more as additional_index_keys)

I'm not sure how to ensure those things are present without either doing what we're doing, or listing a full hardcoded list of the keys we always want, which doesn't seem particularly less hacky than dynamically generating such a list.

The other alternative, it seems, would just be removing the guardrails altogether and just requiring the user to pass in reasonable things to index_keys, which seems worrying because the logic of what index_keys is doing is so deep in the codebase.

File: src/modelfree/score_agent.py (hunk 23 above)
Comment (decodyng) [reply src/modelfree/score_agent.py]: This isn't needed anymore; I'll remove it

Comment (decodyng) [reply src/modelfree/score_agent.py]: Yeah, I'm not really sure what to do here. The reason for this existing is that masking_type is within a dictionary rather than a top level parameter. I don't really like the pattern of making every argument we might pass to the mask_agent class a top level parameter, but I guess it might be the lesser of two evils here. (And, yes, the reason for this was that the kwargs dictionary wasn't hashable)

File: src/modelfree/score_agent.py (hunk 24 above)
Comment (decodyng) [reply src/modelfree/score_agent.py]: Would the idea be to keep the NoisyAgentWrapper in the file with the policies, or with the wrappers? It seems to conceptually fit more with the wrappers, but inherits from DummyModel

File: experiments/modelfree/plot_noisy_observations.py (hunk 8 above)
Comment (decodyng) [reply experiments/modelfree/plot_noisy_observations.py]: I refactored to both the experiment directory and an out-directory to be within /data/aws, but it seems probable that the way I've done it will go against what you want in those directories somehow, in which case I can change it.

File: experiments/modelfree/plot_noisy_observations.py (hunk 14 above)
Comment (decodyng) [reply experiments/modelfree/plot_noisy_observations.py]: These are the current experiments I have run; this is just an example usage, and which experiments you want to visualize will be a function of which you have most recently run, which I don't know how I would anticipate

File: experiments/modelfree/noisy_actions_and_obs.sh (hunk 1 above)
Comment (decodyng) [reply experiments/modelfree/noisy_actions_and_obs.sh]: Done

File: experiments/modelfree/noisy_actions_and_obs.sh (hunk 7 above)
Comment (decodyng) [reply experiments/modelfree/noisy_actions_and_obs.sh]: I do worry about making this more difficult to understand, since I find Bash fairly cryptic to read. I can give refactoring a try when I come back to this over the weekend.

File: src/modelfree/score_agent.py (hunk 24 above)
Comment (AdamGleave) [reply src/modelfree/score_agent.py]: Tricky to communicate so I've moved stuff around to try and address this. Please take a look and let me know if you spot any issues.

File: experiments/modelfree/plot_noisy_observations.py (hunk 14 above)
Comment (decodyng) [reply experiments/modelfree/plot_noisy_observations.py]: Okay, sure. The reason I was putting it in main was that I think it's useful to have an example of "here are directories you can use to have a version this code that is known to work", but I can just remove that if it doesn't seem useful.

File: src/modelfree/configs/multi/score.py (hunk 15 above)
Comment (decodyng) [reply src/modelfree/configs/multi/score.py]: Yeah, I guess; it seems in general worryingly fragile to have configs that only work if you run them in a certain order we'd have to document somewhere. It seems like the most sensible thing to do would be to have the base config be something like an empty string, and then add to it with each named config, but that gets broken by Sacred's  decision about the order of running named vs not-named configs.

File: src/modelfree/configs/ray/aws-noisy-actions-observations.yaml (hunk 21 above)
Comment (decodyng) [reply src/modelfree/configs/ray/aws-noisy-actions-observations.yaml]: Having to search back through Git in order to get an experiment to run seems like not what we should be aiming for, IMO. I'd be potentially fine documenting it if it's documented in a way that it's impossible to avoid seeing if you are running the experiments in question (i.e. directly referenced in the bash script somehow). It seems like probably the best MVP solution is to have a subfolder in ray configs that basically just holds the lines that need to be added, and the anchor lines that need to come before or after them, but trying to build an entire full-featured dynamic YAML merger just for this one use case seems probably like overkill?

I basically just think that anything where you're relying on "this one weird trick you need to run this experiment" living in oral history inside someone's head, that's a bad situation to be in.

File: src/modelfree/multi/common.py [hunk 27]
```diff
@@ -126,7 +126,8 @@ def local_config(platform, local_dir, spec):
             platform = 'local'
 
         if platform == 'local':
-            spec['sync_function'] = 'mkdir -p {remote_dir} && cp -aTv {local_dir}/ {remote_dir}'
+            spec['sync_function'] = 'mkdir -p {remote_dir} && cp -aTv ' \
```
[the remark this answers is not in the document]
Comment (decodyng) [reply src/modelfree/multi/common.py]: This was just what Pycharm's default is, but, fine, can change

File: src/modelfree/configs/multi/score.py (hunk 15 above)
Comment (AdamGleave) [reply src/modelfree/configs/multi/score.py]: I agree this is not ideal. However, the current implementation is also sensitive to the order in which you run things. If you do `medium_accuracy` last, it will clobber the experiment name. Better to at least be consistent with the existing code.

There might be some way of detecting if exp_name was already set? Maybe by a default parameter?
